flight/PiOS/STM32F30x/pios_ws2811.c

```diff
@@ -0,0 +1,226 @@
+/**
+ ******************************************************************************
+ *
+ * @file       pios_ws2811.c
+ * @author     Cleanflight/Betaflight
+ * @author     The LibrePilot Project, http://www.librepilot.org Copyright (C) 2017.
+ * @author     dRonin, http://dRonin.org/, Copyright (C) 2017
+ * @brief      A driver for ws2811 rgb led controller.
+ *             this is a port of the CleanFlight/BetaFlight implementation,
+ *             and in turn ported from LibrePilot.
+ * @see        The GNU Public License (GPL) Version 3
+ *
+ *****************************************************************************/
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ */
+
+#include "pios.h"
+#include "pios_ws2811.h"
+
+#include "pios_dma.h"
+
+#define WS2811_BITS_PER_LED        24
+// for 50us delay
+#define WS2811_DELAY_BUFFER_LENGTH 42
+
+#define WS2811_TIMER_HZ            24000000
+#define WS2811_TIMER_PERIOD        29
+// timer compare value for logical 1
+#define BIT_COMPARE_1              17
+// timer compare value for logical 0
+#define BIT_COMPARE_0              9
+
+#define PIOS_WS2811_MAGIC 0x00281100
+
+struct ws2811_dev_s {
+	uint32_t magic;
+	const struct pios_ws2811_cfg *config;
+
+	bool dma_active;
+
+	int max_leds;
+	int buffer_size;
+
+	uint8_t dma_buffer[];
+};
+
+int PIOS_WS2811_init(ws2811_dev_t *dev_out,
+		const struct pios_ws2811_cfg *ws2811_cfg, int max_leds)
+{
+	if (max_leds <= 0) {
+		return -2;
+	}
+
+	int buffer_size = WS2811_BITS_PER_LED * max_leds +
+			WS2811_DELAY_BUFFER_LENGTH;
+
+	ws2811_dev_t ws2811_dev = PIOS_malloc(sizeof(*ws2811_dev) + buffer_size);
+
+	if (!ws2811_dev) {
+		return -3;
+	}
+
+	memset(ws2811_dev, 0, sizeof(*ws2811_dev) + buffer_size);
+
+	ws2811_dev->magic = PIOS_WS2811_MAGIC;
+	ws2811_dev->config = ws2811_cfg;
+	ws2811_dev->buffer_size = buffer_size;
+	ws2811_dev->max_leds = max_leds;
+
+	TIM_TimeBaseInitTypeDef TIM_TimeBaseStructure;
+	TIM_OCInitTypeDef TIM_OCInitStructure;
+	DMA_InitTypeDef DMA_InitStructure;
+
+	GPIO_InitTypeDef gpio_cfg = {
+		.GPIO_Pin = ws2811_cfg->gpio_pin,
+		.GPIO_Mode = GPIO_Mode_AF,
+		/* Drive hard-- the pin is loaded on F3E, and there could be
```

Comment:
Yah, but the reason I sped it up was the loading on F3E, figuring it was harmless for the rest of targets.  Seems OK to have the comment here (else someone might test a lower speed on e.g. sparky1 and like it more, and think this could change).